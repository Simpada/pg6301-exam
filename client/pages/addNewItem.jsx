import {useContext, useState} from "react";
import {RestaurantApiContext} from "../restaurantApiContext";
import {useNavigate} from "react-router-dom";

export function AddNewItem() {

    const {addItem} = useContext(RestaurantApiContext);
    const [name, setName] = useState("");
    const [ingredients, setIngredients] = useState("");
    const [price, setPrice] = useState("");
    const [category, setCategory] = useState("");

    const navigate = useNavigate();

    function handleSubmit(e) {
        e.preventDefault();
        addItem({name, ingredients: ingredients.split(" "), price: parseInt(price), category});
        navigate("/");
    }

    return(
        <form onSubmit={handleSubmit}>
            <h1> Add new Item to The Menu </h1>

            <div className="form-input">
                <label>
                    <strong>Dish Name</strong>{" "}
                    <input value={name} onChange={(e) => setName(e.target.value)} />
                </label>
            </div>
            <div className="form-input">
                <label>

                    <strong>Ingredients</strong>{" "}
                    <input value={ingredients} onChange={(e) => setIngredients(e.target.value)} />
                    Separate the ingredients by spaces
                </label>
            </div>
            <div className="form-input">
                <label>
                    <strong>Price</strong>{" "}
                    <input
                        onKeyPress={(event) => {
                        if (!/[0-9]/.test(event.key)) {
                            event.preventDefault();
                        }}}
                        value={price} onChange={(e) => setPrice(e.target.value)} />
                </label>
            </div>
            <div className="form-input">
                <label>
                    <strong>Category</strong>{" "}
                    <input value={category} onChange={(e) => setCategory(e.target.value)} />
                </label>
            </div>
            <div>
                <button>Submit</button>
            </div>
        </form>
    )

}