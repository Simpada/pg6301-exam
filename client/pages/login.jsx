import { Link, useNavigate } from "react-router-dom";
import React, { useContext, useState } from "react";
import { RestaurantApiContext } from "../restaurantApiContext";

export function LoginLinks() {
  return (
    <div>
      <div>
        <Link to={"/login"}>Login</Link>
      </div>
      <div>
        <Link to={"/register"}>Register new user</Link>
      </div>
    </div>
  );
}

export function Login() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  const navigate = useNavigate();

  const { login } = useContext(RestaurantApiContext);

  async function handleSubmit(e) {
    e.preventDefault();
    await login({ username, password });
    navigate("/");
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Please log in</h1>
      <div>
        Username:{" "}
        <input value={username} onChange={(e) => setUsername(e.target.value)} />
      </div>
      <div>
        Password:{" "}
        <input
          type="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
      </div>
      <div>
        <button>Log in</button>
      </div>
      <pre>{JSON.stringify({ username, password }, undefined, "  ")}</pre>
    </form>
  );
}
