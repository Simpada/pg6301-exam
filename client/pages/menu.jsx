import {useLoading} from "../useLoading";
import React, {useContext, useState} from "react";
import {RestaurantApiContext} from "../restaurantApiContext";


function ItemEntry({item: {name, ingredients, price, category}}) {
    return <><h3>{name}</h3>
        <div>Ingredients: {ingredients.map(e => `${e}, `)}</div>
        <div>Price: {price}</div>
        <div>Category: {category}</div>
    </>
}

export function Menu() {

    const {menu} = useContext(RestaurantApiContext)
    const [category, setCategory] = useState("");
    const [categoryQuery, setCategoryQuery] = useState("");
    const [price, setPrice] = useState("");
    const [priceQuery, setPriceQuery] = useState("");
    const {loading, error, data} = useLoading(
        async () => await menu({category, price}),
        [category, price]
    );


    function handleSubmitQuery(e) {
        e.preventDefault();
        setCategory(categoryQuery);
        setPrice(priceQuery);
    }


    if (loading){
        return <div>Loading...</div>
    }
    if (error) {
        return <div>
            <h1>ERROR</h1>
            <div id="error-text">{error.toString()}</div>
        </div>
    }

    return <div>
        <h1> The Grand Menu </h1>

        <div>
            <form onSubmit={handleSubmitQuery}>
                <label>
                    Category:
                    <input
                        id="category-query"
                        value={categoryQuery}
                        onChange={(e) => setCategoryQuery(e.target.value)}
                    />
                </label>
                <br/>
                <label>
                    Price:
                    <input
                        id="price-query"
                        onKeyPress={(event) => {
                            if (!/[0-9]/.test(event.key)) {
                                event.preventDefault();
                            }
                        }}
                        value={priceQuery}
                        onChange={(e) => setPriceQuery(e.target.value)}
                    />
                </label>
                <br/>
                <button>Filter</button>
            </form>
        </div>

        {data.map((item) => (
            <ItemEntry key={item.name} item={item}/>
        ))}

    </div>;

}