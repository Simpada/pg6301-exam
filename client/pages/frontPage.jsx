import { Link } from "react-router-dom";
import React, { useContext } from "react";
import { useLoading } from "../useLoading";
import { LoginLinks } from "./login";
import { RestaurantApiContext } from "../restaurantApiContext";

export function FrontPage() {
  const { fetchLogin } = useContext(RestaurantApiContext);
  const { logout } = useContext(RestaurantApiContext);
  const { loading, error, data, reload } = useLoading(
    async () => await fetchLogin()
  );
  const user = data;

  console.log(user);

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return <div>{error.toString()}</div>;
  }

  let admin = false;
  if (user !== null) {
    if (user.isAdmin) {
      admin = true;
    }
  }

  return (
    <div>
      <h1>The API and Web-design Restaurant</h1>
      <ul>
        <li>
          <Link to={"/menu"}>View Menu</Link>
        </li>
        {admin ? (
          <li>
            <Link to={"/menu/add"}>Add an Item to The Menu</Link>
          </li>
        ) : (
          <div></div>
        )}
      </ul>

      <div>
        {user ? (
          <div>
            {user.fullName} ({user.username})
            <button
              onClick={async () => {
                await logout();
                await reload();
              }}
            >
              Logout
            </button>
          </div>
        ) : (
          <LoginLinks />
        )}
      </div>
    </div>
  );
}
