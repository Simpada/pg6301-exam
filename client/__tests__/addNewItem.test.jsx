import {MemoryRouter} from "react-router-dom";
import {AddNewItem} from "../pages/addNewItem";
import * as ReactDOM from "react-dom";
import React from "react";


describe("add new item component", () => {
    it("shows item form", () => {
        const element = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter>
                <AddNewItem/>
            </MemoryRouter>,
            element
        );
        expect(element.innerHTML).toMatchSnapshot();
        expect(
            Array.from(element.querySelectorAll("form label strong")).map(
                (e) => e.innerHTML
            )
        ).toEqual(["Dish Name", "Ingredients", "Price", "Category"]);
    });
});