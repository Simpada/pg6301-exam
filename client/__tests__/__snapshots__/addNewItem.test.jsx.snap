// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add new item component shows item form 1`] = `"<form><h1> Add new Item to The Menu </h1><div class="form-input"><label><strong>Dish Name</strong> <input value=""></label></div><div class="form-input"><label><strong>Ingredients</strong> <input value="">Separate the ingredients by spaces</label></div><div class="form-input"><label><strong>Price</strong> <input value=""></label></div><div class="form-input"><label><strong>Category</strong> <input value=""></label></div><div><button>Submit</button></div></form>"`;
